p2p/discovery.go

```diff
@@ -0,0 +1,204 @@
+			pd.logger.Error().Err(err).Msgf("Failed to connect to bootstrap peer %s", pinfo.ID)
+			continue
+		}
+		pd.addPeer(pinfo)
+	}
+
+	//before periodic gossip, start two rounds of warmup; this is to ensure keygen/keysign unit test
+	// success where there might not be enough time for gossip to propagate before the keygen starts.
+	pd.gossipPeers(ctx)
+	time.Sleep(1 * time.Second)
+	pd.gossipPeers(ctx)
+	// Start periodic gossip
+	go pd.startGossip(ctx)
+}
+
+func (pd *PeerDiscovery) Stop() {
+	close(pd.closeChan)
+}
+
+// addPeer adds a peer to known peers
+func (pd *PeerDiscovery) addPeer(pinfo peer.AddrInfo) {
+	pd.mu.Lock()
+	defer pd.mu.Unlock()
+
+	if pinfo.ID == pd.host.ID() {
+		return // Don't add ourselves
+	}
+	pd.knownPeers[pinfo.ID] = pinfo
+}
+
+// GetPeers returns all known peers
+func (pd *PeerDiscovery) GetPeers() []peer.AddrInfo {
+	pd.mu.RLock()
+	defer pd.mu.RUnlock()
+
+	peers := make([]peer.AddrInfo, 0, len(pd.knownPeers))
+	for _, p := range pd.knownPeers {
+		peers = append(peers, p)
+	}
+	return peers
+}
+
+// handleDiscovery handles incoming discovery streams
+func (pd *PeerDiscovery) handleDiscovery(s network.Stream) {
+	pd.logger.Debug().Msgf("Received discovery stream from %s", s.Conn().RemotePeer())
+	defer s.Close()
+
+	ma := s.Conn().RemoteMultiaddr()
+
+	ai := peer.AddrInfo{
+		ID:    s.Conn().RemotePeer(),
+		Addrs: []multiaddr.Multiaddr{ma},
+	}
+	pd.addPeer(ai)
+
+	// Share our known peers
+	peers := pd.GetPeers()
+	data, err := json.Marshal(peers)
+	if err != nil {
+		pd.logger.Error().Err(err).Msgf("Failed to marshal peers")
+		return
+	}
+	_, err = s.Write(data)
+	if err != nil {
+		pd.logger.Error().Err(err).Msgf("Failed to write to stream")
+	}
+}
+
+// startGossip periodically shares peer information
+func (pd *PeerDiscovery) startGossip(ctx context.Context) {
+	ticker := time.NewTicker(GossipInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case _, ok := <-pd.closeChan:
+			if !ok {
+				pd.logger.Info().Msg("Peer discovery stopped")
+				return
+			}
+			pd.logger.Warn().Msgf("Should not receive from closed channel!")
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			pd.gossipPeers(ctx)
+		}
+	}
+}
+
+func (pd *PeerDiscovery) gossipPeers(ctx context.Context) {
+	pd.logger.Debug().Msgf("Gossiping known peers")
+	peers := pd.GetPeers()
+	pd.logger.Debug().Msgf("current peers: %v", peers)
+
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+	var wg sync.WaitGroup
+	sem := make(chan struct{}, MaxGossipConcurrency) // Limit concurrency
+
+	for _, p := range peers {
+		if p.ID == pd.host.ID() {
+			continue
+		}
+
+		sem <- struct{}{}
+		wg.Add(1)
+		go func(p peer.AddrInfo) {
+			defer wg.Done()
+			defer func() { <-sem }()
+
+			err := pd.host.Connect(ctx, p)
+			if err != nil {
+				pd.logger.Error().Err(err).Msgf("Failed to connect to peer %s", p)
+				return
+			}
+			pd.logger.Debug().Msgf("Connected to peer %s", p)
```

Comment:
`NewStream()` will connect automatically if needed so I don't think this is needed?

Reply:
yes but i guess it does not hurt--plus the error message might be easier to track.